File: packages/shared/src/types/routine.ts
export type RoutineFrequency = 'daily' | 'weekly' | 'monthly';

export interface RoutineSchedule {
  frequency: RoutineFrequency;
  daysOfWeek?: number[]; // 0-6 for Sun-Sat
  dayOfMonth?: number; // 1-31
  time?: string; // "HH:mm" format
}

export interface Routine {
  id: string;
  coupleId: string;
  name: string;
  description?: string;
  schedule: RoutineSchedule;
  assignedToUserId?: string;
  isActive: boolean;
  createdById: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface RoutineOccurrence {
  id: string;
  routineId: string;
  coupleId: string;
  scheduledDate: Date;
  completedAt?: Date;
  completedById?: string;
  skipped: boolean;
  createdAt: Date;
}

export interface RoutineWithOccurrences extends Routine {
  upcomingOccurrences: RoutineOccurrence[];
  completionRate?: number;
  currentStreak?: number;
}

export interface RoutineOccurrenceWithDetails extends RoutineOccurrence {
  routine: Routine;
  completedBy?: {
    id: string;
    displayName: string;
    avatarUrl?: string;
  };
}

File: apps/ios/TwoDo/Models/Routine.swift
import Foundation

enum RoutineFrequency: String, Codable {
    case daily
    case weekly
    case monthly
}

struct RoutineSchedule: Codable {
    var frequency: RoutineFrequency
    var daysOfWeek: [Int]? // 0 = Sunday, 6 = Saturday
    var dayOfMonth: Int? // 1-31

    // Helper to check if routine is due on a given date
    func isDueOn(_ date: Date) -> Bool {
        let calendar = Calendar.current
        let weekday = calendar.component(.weekday, from: date) - 1 // Convert to 0-6
        let day = calendar.component(.day, from: date)

        switch frequency {
        case .daily:
            return true
        case .weekly:
            return daysOfWeek?.contains(weekday) ?? false
        case .monthly:
            return dayOfMonth == day
        }
    }
}

struct Routine: Codable, Identifiable {
    let id: String
    let coupleId: String
    var name: String
    var description: String?
    var schedule: RoutineSchedule
    var assignedToUserId: String?
    let createdById: String
    let createdAt: Date
    let updatedAt: Date
}

struct RoutineOccurrence: Codable, Identifiable {
    let id: String
    let routineId: String
    let scheduledDate: Date
    var completedAt: Date?
    var completedById: String?
    var skipped: Bool
    let createdAt: Date
    let updatedAt: Date
}

struct RoutineWithOccurrences: Codable {
    let routine: Routine
    let occurrences: [RoutineOccurrence]
}

struct RoutineStats: Codable {
    var currentStreak: Int
    var longestStreak: Int
    var totalCompleted: Int
    var completionRate: Double
}

// MARK: - Routine Request/Response Models

struct GetRoutinesResponse: Codable {
    let routines: [Routine]
}

struct CreateRoutineRequest: Codable {
    let name: String
    let description: String?
    let schedule: RoutineSchedule
    let assignedToUserId: String?
}

struct CreateRoutineResponse: Codable {
    let routine: Routine
}

struct GetRoutineResponse: Codable {
    let routine: Routine
    let occurrences: [RoutineOccurrence]
}

struct UpdateRoutineRequest: Codable {
    let name: String?
    let description: String?
    let schedule: RoutineSchedule?
    let assignedToUserId: String?
}

struct UpdateRoutineResponse: Codable {
    let routine: Routine
}

struct CompleteOccurrenceResponse: Codable {
    let occurrence: RoutineOccurrence
}

struct SkipOccurrenceResponse: Codable {
    let occurrence: RoutineOccurrence
}

struct GetRoutineStatsResponse: Codable {
    let stats: RoutineStats
}
